x/mongo/driver/operation.go

```diff
@@ -1468,15 +1526,18 @@ func (op Operation) addWriteConcern(dst []byte, desc description.SelectedServer)
 		return dst, nil
 	}
 
-	t, data, err := wc.MarshalBSONValue()
-	if errors.Is(err, writeconcern.ErrEmptyWriteConcern) {
+	wtimeout := getWriteConcernTimeout(ctx, op)
+
+	typ, wcBSON, err := marshalBSONWriteConcern(*wc, wtimeout)
```

Comment:
I interpret ["[wTimeoutMS] option[s] MUST be deprecated in favor of timeoutMS"](https://github.com/mongodb/specifications/blob/master/source/client-side-operations-timeout/client-side-operations-timeout.md#deprecations) to mean that if a user sets a client-level timeout, then that value should be applied to the `wTimeoutMS`. Otherwise, this condition holds when [committing a transaction](https://github.com/mongodb/specifications/blob/ac903bf6edb859456c1005a439efcd0769e10870/source/transactions/transactions.rst#L441-L451): 

> if the modified write concern does not include a wtimeout value, drivers MUST also apply wtimeout: 10000 to the write concern in order to avoid waiting forever (or until a socket timeout) if the majority write concern cannot be satisfied.

Reply:
I interpreted the section you referenced as describing how `timeoutMS` replaces those other driver-side configuration values. The driver configuration values and database command fields are often identical, so there is some vagueness in the CSOT spec. In any case, setting `maxTimeMS` should have the same effect as `wTimeout` in the case that the operation is blocked on satisfying the write concern, so setting `wTimeout` in commands seems unnecessary.

I think that section of the Transactions spec is probably vestigial since the scenario it describes is fully covered by `timeoutMS` (i.e. if you set a client-side timeout, an operation will never wait forever).

Reply:
@ShaneHarvey Do you have any thoughts on whether or not CSOT should be applied to `wTimeout` ? This closely mirrors the 1.x issue where the context deadline does not get applied to `maxTimeMS`.

Reply:
> I think that section of the Transactions spec is probably vestigial since the scenario it describes is fully covered by timeoutMS (i.e. if you set a client-side timeout, an operation will never wait forever).

@matthewdale I think I follow, are you suggesting that we retain this logic but only if there is no CSOT? 

```go
func getWriteConcernTimeout(ctx context.Context, op Operation) time.Duration {
	if contextDeadline, ok := ctx.Deadline(); ok {
		return 0
	}

	if op.Timeout != nil {
		return 0
	}

	if op.Client == nil {
		return 0
	}

	return op.Client.CurrentWTimeout
}
```

Reply:
CSOT does not set wTimeout. The server already applies maxTimeMS to the entire command including the wait for write concern step so setting wTimeout is not needed.

Reply:
For example see: https://github.com/mongodb/specifications/blob/e9ae09d/source/client-side-operations-timeout/tests/deprecated-options.json#L142

Reply:
It's difficult to tell from the deprecated-options unified spec test, since those options won't exist in 2.0. To be clear, the CSOT proposed in this PR removes wtimeout from the WriteConcern options in the stable API.

Reply:
Do we need a new spec test that asserts wTimeout is not set and maxTimeMS is?

Reply:
@ShaneHarvey That would be helpful. Should I open a ticket for that?

@matthewdale Concerning this:

> I think that section of the Transactions spec is probably vestigial since the scenario it describes is fully covered by timeoutMS

What happens in the case where a user does not set a CSOT? We would be spec non-compliant and the transaction could wait indefinitely. Additionally, here is a list of tests legacy tests that we would need to skip, all of which check the default value of 10s for wtimeout during a transaction:

```
TestUnifiedSpecs/transactions/legacy/error-labels.json/add_RetryableWriteError_and_UnknownTransactionCommitResult_labels_to_connection_errors

TestUnifiedSpecs/transactions/legacy/error-labels.json/add_RetryableWriteError_and_UnknownTransactionCommitResult_labels_to_retryable_commit_errors

TestUnifiedSpecs/transactions/legacy/error-labels.json/add_RetryableWriteError_and_UnknownTransactionCommitResult_labels_to_writeConcernError_ShutdownInProgress

TestUnifiedSpecs/transactions/legacy/error-labels.json/add_UnknownTransactionCommitResult_label_to_writeConcernError_WriteConcernFailed

TestUnifiedSpecs/transactions/legacy/error-labels.json/add_UnknownTransactionCommitResult_label_to_writeConcernError_WriteConcernFailed_with_wtimeout

TestUnifiedSpecs/transactions/legacy/retryable-commit-errorLabels.json/commitTransaction_retries_once_with_RetryableWriteError_from_server

TestUnifiedSpecs/transactions/legacy/retryable-commit.json/commitTransaction_fails_after_two_errors

TestUnifiedSpecs/transactions/legacy/retryable-commit.json/commitTransaction_succeeds_after_connection_error

TestUnifiedSpecs/transactions/legacy/retryable-commit.json/commitTransaction_succeeds_after_NotWritablePrimary

TestUnifiedSpecs/transactions/legacy/retryable-commit.json/commitTransaction_succeeds_after_NotPrimaryOrSecondary

TestUnifiedSpecs/transactions/legacy/retryable-commit.json/commitTransaction_succeeds_after_NotPrimaryNoSecondaryOk

TestUnifiedSpecs/transactions/legacy/retryable-commit.json/commitTransaction_succeeds_after_InterruptedDueToReplStateChange

TestUnifiedSpecs/transactions/legacy/retryable-commit.json/commitTransaction_succeeds_after_InterruptedAtShutdown

TestUnifiedSpecs/transactions/legacy/retryable-commit.json/commitTransaction_succeeds_after_PrimarySteppedDown

TestUnifiedSpecs/transactions/legacy/retryable-commit.json/commitTransaction_succeeds_after_ShutdownInProgress

TestUnifiedSpecs/transactions/legacy/retryable-commit.json/commitTransaction_succeeds_after_HostNotFound

TestUnifiedSpecs/transactions/legacy/retryable-commit.json/commitTransaction_succeeds_after_HostUnreachable

TestUnifiedSpecs/transactions/legacy/retryable-commit.json/commitTransaction_succeeds_after_SocketException

TestUnifiedSpecs/transactions/legacy/retryable-commit.json/commitTransaction_succeeds_after_NetworkTimeout

TestUnifiedSpecs/transactions/legacy/retryable-commit.json/commitTransaction_succeeds_after_WriteConcernError_InterruptedAtShutdown

TestUnifiedSpecs/transactions/legacy/retryable-commit.json/commitTransaction_succeeds_after_WriteConcernError_InterruptedDueToReplStateChange

TestUnifiedSpecs/transactions/legacy/retryable-commit.json/commitTransaction_succeeds_after_WriteConcernError_PrimarySteppedDown

TestUnifiedSpecs/transactions/legacy/retryable-commit.json/commitTransaction_succeeds_after_WriteConcernError_ShutdownInProgress

TestUnifiedSpecs/convenient-transactions/commit-retry.json/commitTransaction_succeeds_after_multiple_connection_errors

TestUnifiedSpecs/convenient-transactions/commit-retry.json/commit_is_retried_after_commitTransaction_UnknownTransactionCommitResult_(NotWritablePrimary)

TestUnifiedSpecs/convenient-transactions/commit-writeconcernerror.json/commitTransaction_is_retried_after_WriteConcernFailed_timeout_error

TestUnifiedSpecs/convenient-transactions/commit-writeconcernerror.json/commitTransaction_is_retried_after_WriteConcernFailed_non-timeout_error
```

Reply:
Reading the referenced [commitTransaction](https://github.com/mongodb/specifications/blob/master/source/transactions/transactions.md#committransaction) section of the Transactions spec and the referenced spec tests, we're talking specifically about the case when the driver retries `commitTransaction` and overrides some fields on the write concern. 

In that case, there are two questions to answer:
1. If `timeoutMS` is not set at any level, should drivers include the default `wTimeout` value of 10,000 on retried `commitTransaction` commands?
2. If `timeoutMS` is set (at any level), should drivers include a `wTimeout` value on retried `commitTransaction` commands? If so, should the value be the same as `maxTimeMS`?

These are my proposed answers:
1. That seems like a good idea. It maintains the existing default behavior and doesn't seem to conflict with CSOT because there is no CSOT timeout specified.
2. In that case, `wTimeout` would be redundant since the command should include `maxTimeMS`. It seems safe to omit `wTimeout`.

TLDR: I think we should include `wTimeout: 10000` on retried `commitTransaction` commands if `timeoutMS` is not set, and omit `wTimeout` if `timeoutMS` is set. That behavior should also satisfy all of the referenced spec tests.